L64-python-runtime-proto-parsing.md

```diff
@@ -0,0 +1,99 @@
+
+```python
+protos = grpc.protos('foo.proto')
+services = grpc.services('foo.proto')
+```
+
+These two new functions return the same module objects as the `import foo_pb2` and `import foo_pb2_grpc` statements. In order to maintain interoperability with any proto-backed modules loaded in the same process, after these functions are invoked, `import foo_pb2` and `import foo_pb2_grpc` will be no-ops. That is, a side-effect of calling `grpc.protos` and `grpc.services` is insertion of the returned modules into the per-process cache. This ensures that regardless of whether the application calls `grpc.protos('foo.proto')` or `import foo_pb2` or in which order, only a single version of the module will ever be loaded into the process. This avoids situations in which interoperability breaks due to two modules expecting the same protobuf-level message type, but expecting two *different* Python-level `Message` classes, one backed by a `_pb2.py` file and one backed by a `.proto` file.
+
+The wrapper function around the import serves several purposes here. First, it puts the user in control of naming the module (in a manner similar to Javascript), meaning the user never has to concern themself with the confusing `_pb2` suffix. Second, the function provides a wrapping layer through which the library can provide guidance in the case of failed imports.
+
+To be precise, we propose the introduction of three new functions with the following signatures:
+
+```python
+def protos(proto_file: Text,
+           runtime_generated: Optional[bool] = True) -> types.ModuleType:
+    pass
+
+def services(proto_file: Text,
+             runtime_generated: Optional[bool] = True) -> types.ModuleType:
+    pass
+
+def protos_and_services(proto_file: Text,
+                        runtime_generated: Optional[bool] = True) -> Tuple[types.ModuleType, types.ModuleType]:
+    pass
+```
+
+The final function, `protos_and_services` is a simple convenience function allowing the user to import protos and services in a single function call.
+
+The change will be entirely backward compatible. Users manually invoking `protoc` today will not be required to change their build process.
+
+### Import Paths
+
+These functions will behave like normal `import` statements. `sys.path` will be used to search for `.proto` files. The path under which each particular `.proto` file was found will be passed to the protobuf parser as the root of the tree (equivalent to the `-I` flag of `protoc`). This means that a file located at `${SYS_PATH_ENTRY}/foo/bar/baz.proto` will result in the instantation of a module with the fully qualified name `foo.bar.baz_pb2`.  Users are expected to have a directory structure mirroring their desired import structure.
+
+Users have reported that getting Protobuf paths to match Python import paths is quite tricky today using `protoc`. In the case of a failure to import, the library will print troubleshooting information alongside the error.
```

Comment:
In the implementation, I recall you try to make the import path identical to the package name. Is it still true, should we mention it?

Reply:
I'm not sure what you're referring to here. Unfortunately, both Python and Protobuf use the terms "package" and "import", so there are actually four things to worry about for every module. As written right now, the Python package, Python import path, and Protobuf import path all match up, while the Protobuf package is entirely independent.

So, taking the `grpc/grpc` repo as an example, supposing the repo root were on your `PYTHONPATH` (which would be the case if the interpreter were started from the repo root), [test.proto](https://github.com/grpc/grpc/blob/master/src/proto/grpc/testing/test.proto) would be importable with `grpc.protos('src/proto/grpc/testing/test.proto')` or, equivalently, `src.proto.grpc.testing.test_pb2`. Notice that, within the proto file itself, it's importing other files with the "src/proto" prefix. This is the Protobuf-level import path. This means that the repo root *must* be on the protocol buffer parser's search path, otherwise, this file will fail to import its dependencies.

The Protobuf package name for this file is "grpc.testing", which is much more palatable, since it really is just a namespacing mechanism and doesn't let the repo organization directories leak through. This has no bearing on how the module is imported from within Python. It *is* very common, however, to have the package at least be a suffix of the Protobuf import path, as is the case here. In that case, (in theory -- I haven't tested this), you should be able to add both the repo root and "src/proto" to your `PYTHONPATH` and be able to import you modules with `grpc.protos('grpc/testing/test.proto')` or `import grpc.testing.test_pb2`.

In general, I would recommend that people not include anything like the "src/proto" prefix in their ".proto" files at all. There should be some directory within their repo that they expect to be the root of their proto source tree and all Protobuf-level `import` statements should be made relative to *that*. That is, [this line](https://github.com/grpc/grpc/blob/e677f3d166ebcac000e2657263b3505c5c2f701f/src/proto/grpc/testing/test.proto#L21) should be changed to `import "grpc/testing/empty.proto"`. This would get all four attributes to line up and would ensure that the user only ever has to add that unique directory to their `PYTHONPATH`. Or, equivalently, to start the interpreter from that directory.

I'll add a section to the document explaining this.

Reply:
Make the import path look like the Protobuf package name is much better than the arbitrary path created by users. The explanation is very informative, people who read this design doc will appreciate this rationale.